### PR TITLE
Standardises item blending

### LINKED ISSUE #22024: Juicing Plump Helmets Destroys them
(WEB REPORT BY: turingwept REMOTE: 206.221.180.138:7777)
# Revision
bdd516573cc6f178b6e13b2f76dd7494fb98adbd

# Description
Normally when an object is juiced it either produces a juice reagent, or it fails to go through and spits the item back out.  This one did neither, destroying the object and then producing nothing.
# Steps to Reproduce

Put Plump Helmet into the Juicer, Juice it. 
# What you Expected

Plump Helmet Wine if it was implemented yet.  Otherwise nothing.
# What Actually Happened
Destroyed the plump helmet

### LINKED ISSUE #34237: Grinding Death Nettles is broken
Death nettles with modified phytochemistry genes seem to be incredibly fucked up.
I just witnessed a botanist creating nettles with 8.2 units of valerenic acid and 40.8 units of plasticide.
Grinding those samples in the all-in-one-grinder net NOTHING, not a single reagent. It was simply gone.
Grinding those samples with a mortar net 10 phenol, the reagents of generic death nettles, despite those samples not having the reagent at all!

Sadly I couldn't do further testing with other chemicals or regular nettles.

### PR DESCRIPTION
**Pull request by SECBATON-GRIFFON**
[bugfix][consistency]

## What this does
takes the blend_items lists of mortars and grinders and combines their values onto vars on types of /obj/item, any values that are higher in the grinder list were kept over the mortar's.
does the same with juice_items.
cuts down code to make it reused by both items.
makes grinding nettles actually give their chems right for both.
Closes #34237.
Closes #22024.

## Why it's good
consistency

## How it was tested
blend_juice unit test was written for most of this, grinder and mortar in-game otherwise

## Changelog
:cl:
 * bugfix: Nettles and death nettles now properly grind their reagents out.
 * tweak: Mortars can now grind anything an all-in-one grinder can, to the same amounts, and vice versa.
 * tweak: Ground novaflowers, nettles and death nettles now grind their reagents in scale with potency.

### CODE REVIEW

**Comment by Eneocho** [comment]
Novaflowers are supposed to give novaflour when ground, so you can bake novabread. 
Would this break that interaction and squeeze the capsaicin inside instead?

**Comment by SECBATON-GRIFFON** [comment]
> Novaflowers are supposed to give novaflour when ground, so you can bake novabread. Would this break that interaction and squeeze the capsaicin inside instead?

thanks for catching that, reverting it to old value

**Comment by Eneocho** [comment]
Actually I'd like it if you could fix how the grinders handle the novaflowers, cause they always spit out 15u novaflour regardless of potency. 
Also, wheat grinds into flour instead of nutriment, does this handle that correctly? Unlike novaflour, the wheat's potency does increase the flour output.

**Comment by SECBATON-GRIFFON** [comment]
> Actually I'd like it if you could fix how the grinders handle the novaflowers, cause they always spit out 15u novaflour regardless of potency.

consider it done

> Also, wheat grinds into flour instead of nutriment, does this handle that correctly? Unlike novaflour, the wheat's potency does increase the flour output.

was always the case in the code

File: code/modules/reagents/machinery/reagentgrinder.dm [hunk 1]
```diff
@@ -399,55 +402,43 @@ var/global/list/juice_items = list (
 		holdingitems -= O
 	holdingitems = list()
 
-/obj/machinery/reagentgrinder/proc/is_allowed(var/obj/item/weapon/reagent_containers/O)
+/proc/is_allowed(var/obj/item/weapon/reagent_containers/O)
```
Comment (jwhitak) [inline code/modules/reagents/machinery/reagentgrinder.dm]: We require absolute type paths in submitted code, so please revert this change (alongside the additional cases in this file)

File: code/modules/reagents/machinery/reagentgrinder.dm [hunk 2]
```diff
@@ -399,55 +402,43 @@ var/global/list/juice_items = list (
...
-
-/obj/machinery/reagentgrinder/proc/get_juice_amount(var/obj/item/weapon/reagent_containers/food/snacks/grown/O)
-	if (!istype(O))
-		return 5
-	else if (O.potency == -1)
-		return 5
-	else
-		return round(5*sqrt(O.potency))
+/proc/get_juice_amount(var/obj/item/weapon/reagent_containers/food/snacks/grown/O)
+	return !istype(O) || O.potency <= 0 ? 5 : round(5*sqrt(O.potency))
```
Comment (jwhitak) [inline code/modules/reagents/machinery/reagentgrinder.dm]: The earlier version with the if/else tree maintains readability compared to this version, you could reduce the tree while keeping things less of a (one line blur) by using something like

```
if (!istype(O) || O.potency <= 0)
    return 5
else
    return round(5*sqrt(O.potency))
````

File: code/modules/reagents/machinery/reagentgrinder.dm [hunk 3]
```diff
@@ -399,55 +402,43 @@ var/global/list/juice_items = list (
...
-		return
-	if (!beaker || (beaker && beaker.reagents.total_volume >= beaker.reagents.maximum_volume))
-		return
-	playsound(src, speed_multiplier < 2 ? 'sound/machines/juicer.ogg' : 'sound/machines/juicerfast.ogg', 30, 1)
-	inuse = 1
-	spawn(50/speed_multiplier)
-		inuse = 0
-		interact(usr)
+/obj/machinery/reagentgrinder/proc/juice(var/test = FALSE)
+	if(!test)
```
Comment (jwhitak) [inline code/modules/reagents/machinery/reagentgrinder.dm]: There has to be a better way than to add a permanent testing argument into this function. Perhaps change up the object/procs so that you can attempt the power/beaker/use checks before calling juice()?

File: code/modules/reagents/machinery/reagentgrinder.dm [hunk 4]
```diff
@@ -469,105 +460,88 @@ var/global/list/juice_items = list (
...
-
-		if(O.reagents.reagent_list.len == 0)
-			remove_object(O)
-
-
-	//Sheets
-	for(var/obj/item/stack/sheet/O in holdingitems)
-		var/allowed = get_allowed_by_id(O)
+/obj/machinery/reagentgrinder/proc/grind(var/test = FALSE)
+	if(!test)
```
Comment (jwhitak) [inline code/modules/reagents/machinery/reagentgrinder.dm]: Same as above

File: code/modules/unit_tests/blend_juice.dm [hunk 5]
```diff
@@ -0,0 +1,40 @@
+/datum/unit_test/grind_juice/start()
```
Comment (jwhitak) [inline code/modules/unit_tests/blend_juice.dm:1]: Unit tests my beloved

File: code/modules/reagents/machinery/reagentgrinder.dm [hunk 6]
```diff
@@ -399,55 +402,43 @@ var/global/list/juice_items = list (
...
 
-/obj/machinery/reagentgrinder/proc/is_allowed(var/obj/item/weapon/reagent_containers/O)
+/proc/is_allowed(var/obj/item/weapon/reagent_containers/O)
 	for (var/i in blend_items)
 		if(istype(O, i))
 			return 1
 	return 0
 
-/obj/machinery/reagentgrinder/proc/get_allowed_by_id(var/obj/item/weapon/grown/O)
+/proc/get_allowed_by_id(var/obj/item/weapon/grown/O)
```
Comment (jwhitak) [inline code/modules/reagents/machinery/reagentgrinder.dm]: Same as above

File: code/modules/reagents/machinery/reagentgrinder.dm [hunk 7]
```diff
@@ -399,55 +402,43 @@ var/global/list/juice_items = list (
...
 	for (var/i in blend_items)
 		if (istype(O, i))
 			return blend_items[i]
 
-/obj/machinery/reagentgrinder/proc/get_allowed_juice_by_id(var/obj/item/weapon/reagent_containers/food/snacks/O)
+/proc/get_allowed_juice_by_id(var/obj/item/weapon/reagent_containers/food/snacks/O)
```
Comment (jwhitak) [inline code/modules/reagents/machinery/reagentgrinder.dm]: Same as above

File: code/modules/reagents/machinery/reagentgrinder.dm [hunk 8]
```diff
@@ -399,55 +402,43 @@ var/global/list/juice_items = list (
...
-		return round(O.potency)
-
-/obj/machinery/reagentgrinder/proc/get_juice_amount(var/obj/item/weapon/reagent_containers/food/snacks/grown/O)
-	if (!istype(O))
-		return 5
-	else if (O.potency == -1)
-		return 5
-	else
-		return round(5*sqrt(O.potency))
+/proc/get_juice_amount(var/obj/item/weapon/reagent_containers/food/snacks/grown/O)
```
Comment (jwhitak) [inline code/modules/reagents/machinery/reagentgrinder.dm]: Same as above

**Review by jwhitak — CHANGES_REQUESTED** [review]

File: code/modules/reagents/machinery/reagentgrinder.dm (hunk 1 above)
Comment (SECBATON-GRIFFON) [reply code/modules/reagents/machinery/reagentgrinder.dm]: removing this proc since it's unused anyways

File: code/modules/reagents/machinery/reagentgrinder.dm (hunk 3 above)
Comment (SECBATON-GRIFFON) [reply code/modules/reagents/machinery/reagentgrinder.dm]: pre_use() function splits this up and cuts down code, done

File: code/modules/reagents/machinery/reagentgrinder.dm (hunk 4 above)
Comment (SECBATON-GRIFFON) [reply code/modules/reagents/machinery/reagentgrinder.dm]: done

File: code/modules/reagents/machinery/reagentgrinder.dm (hunk 2 above)
Comment (SECBATON-GRIFFON) [reply code/modules/reagents/machinery/reagentgrinder.dm]: done

File: code/modules/reagents/machinery/reagentgrinder.dm (hunk 6 above)
Comment (SECBATON-GRIFFON) [reply code/modules/reagents/machinery/reagentgrinder.dm]: reorganised it so the argument is now the thing the proc is called on, if thats any better. mortars use this too so that's my reasoning

File: code/modules/reagents/machinery/reagentgrinder.dm (hunk 7 above)
Comment (SECBATON-GRIFFON) [reply code/modules/reagents/machinery/reagentgrinder.dm]: see above

File: code/modules/reagents/machinery/reagentgrinder.dm (hunk 8 above)
Comment (SECBATON-GRIFFON) [reply code/modules/reagents/machinery/reagentgrinder.dm]: see above

File: code/modules/reagents/machinery/reagentgrinder.dm [hunk 9]
```diff
@@ -399,43 +406,30 @@ var/global/list/juice_items = list (
...
 	holdingitems = list()
 
-/obj/machinery/reagentgrinder/proc/is_allowed(var/obj/item/weapon/reagent_containers/O)
-	for (var/i in blend_items)
-		if(istype(O, i))
-			return 1
-	return 0
-
-/obj/machinery/reagentgrinder/proc/get_allowed_by_id(var/obj/item/weapon/grown/O)
+/obj/item/proc/get_allowed_by_id()
```
Comment (jwhitak) [inline code/modules/reagents/machinery/reagentgrinder.dm]: I was wrong about absolute/relative typepaths before, those were straight up global procs. Reducing their scope to obj/item works. I'd also rename this proc, since "get_allowed_by_id" doesn't really describe this. Get_allowed_blend_by_id would be great.

File: code/modules/reagents/machinery/reagentgrinder.dm [hunk 10]
```diff
@@ -44,13 +29,12 @@ var/global/list/juice_items = list (
...
 		/obj/item/trash/candle				  = list(WAX = 1),
-		/obj/item/weapon/grown/nettle         = list(FORMIC_ACID = 0),
-		/obj/item/weapon/grown/deathnettle    = list(PHENOL = 0),
-		/obj/item/stack/sheet/charcoal        = list("charcoal" = 20),
+		/obj/item/stack/sheet/charcoal        = list(CHARCOAL = 20),
 		/obj/item/stack/sheet/bone	          = list(BONEMARROW = 20),
 
 		//Blender Stuff
-		/obj/item/weapon/reagent_containers/food/snacks/grown/soybeans = list(SOYMILK = -10), //I have no fucking idea what most of these numbers mean and I hate them.
+		//Inverted numbers add these reagents in proportion to the nutriment inside, and minus numbers multiply it, ..--==the more you know!*
```
Comment (jwhitak) [inline code/modules/reagents/machinery/reagentgrinder.dm]: Inverted numbers vs minus numbers? Which adds, which multiplies? There are many confused people over this

**Review by jwhitak — CHANGES_REQUESTED** [review]

**Comment by jwhitak** [comment]
Exxion brought up a good point. Why have this long global list of types and specialty helper procs to determine if something can be ground up from this list? You could include these variables directly on the relevant objects, or have a proc that you can call to return them.

You even have the precursor of this:
```
/obj/item/proc/get_juice_amount()
	return 5
```
With specific subtypes returning different amounts. Now just replace that with an associative list of reagents and how much is contained, with the defaults being just nutriment, and overrides for individual items affected. The upside is that this can allow a juice/grind action to return multiple chemicals.

That said, I can see you not doing any of this because it's pretty beyond the scope of your already powerful bugfix. In fact, Exxion gave me a whole wishlist for a juice/grind rework, including merging the two functions, including the juice in the plants directly, and the blind/grind effect transferring the reagents and possibly additional ones. This is just a standarization PR lol. I'll leave the idea out in the open.

**Comment by Eneocho** [comment]
my two bits on the "including the juice on the plant already" idea:
if you do make it such, make it so the juice depends on the potency and morphology and isn't on the reagents list itself, but rather as an extractable when crushed (so it behaves more or less like it does currently), cause you can splice shit really hard and the reagent cap in fruit is 100, so having a third or more of that cap consumed by the tomato juice is kinda lame. on the other hand we do have bananas, with them having the juice internally already, and you could just purge out the juice while splicing so it just doesn't take space, but that would clash with green grapes and carrots, since those have tannic acid (burn meds) and zeaxanthine (eye meds) respectively.

File: code/modules/reagents/machinery/reagentgrinder.dm (hunk 10 above)
Comment (SECBATON-GRIFFON) [reply code/modules/reagents/machinery/reagentgrinder.dm]: i meant to say zero for inverted, fixed

File: code/modules/reagents/machinery/reagentgrinder.dm (hunk 9 above)
Comment (SECBATON-GRIFFON) [reply code/modules/reagents/machinery/reagentgrinder.dm]: reworked these out anyways

**Comment by SECBATON-GRIFFON** [comment]
> Exxion brought up a good point. Why have this long global list of types and specialty helper procs to determine if something can be ground up from this list? You could include these variables directly on the relevant objects, or have a proc that you can call to return them.
> 
> You even have the precursor of this:
> 
> ```
> /obj/item/proc/get_juice_amount()
> 	return 5
> ```
> 
> With specific subtypes returning different amounts. Now just replace that with an associative list of reagents and how much is contained, with the defaults being just nutriment, and overrides for individual items affected. The upside is that this can allow a juice/grind action to return multiple chemicals.
> 
> That said, I can see you not doing any of this because it's pretty beyond the scope of your already powerful bugfix. In fact, Exxion gave me a whole wishlist for a juice/grind rework, including merging the two functions, including the juice in the plants directly, and the blind/grind effect transferring the reagents and possibly additional ones. This is just a standarization PR lol. I'll leave the idea out in the open.

well, i went ahead and did it anyways, but only made it support a single chemical and amount for now

**Comment by jwhitak** [comment]
Now the hardest part of the PR... updating the wiki 😈
